src/HttpCacheListener.php

```diff
@@ -273,4 +293,70 @@ public function setVary(Headers $headers)
 
         return $this;
     }
+
+    /**
+     * @param HttpRequest $request
+     * @param HttpResponse $response
+     * @return $this
+     */
+    public function setEtag(HttpRequest $request, HttpResponse $response)
+    {
+        $headers = $response->getHeaders();
+        if (! empty($this->cacheConfig['etag'])
+            && (! $headers->has('etag')
+                || ! empty($this->cacheConfig['etag']['override']))
+        ) {
+            $etag = new Header\Etag($this->generateEtag($request, $response));
+            $headers->addHeader($etag);
+        }
+
+        return $this;
+    }
+
+    /**
+     * @param HttpRequest $request
+     * @param HttpResponse $response
+     * @return $this
+     */
+    public function setNotModified(HttpRequest $request, HttpResponse $response)
+    {
+        if (!$request->getHeaders()->has('If-None-Match') || !$response->getHeaders()->has('Etag')) {
+            return $this;
+        }
+
+        $requestEtags = $request->getHeaders()->get('If-None-Match')->getFieldValue();
+        $requestEtags = !is_array($requestEtags) ? [$requestEtags] : $requestEtags;
+        $responseEtag = $response->getHeaders()->get('Etag')->getFieldValue();
+
+        if (in_array($responseEtag, $requestEtags) || in_array('*', $requestEtags)) {
+            $response->setStatusCode(304);
+            $response->setContent(null);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Generates an Etag based on the response.
+     *
+     * @param HttpRequest $request
+     * @param HttpResponse $response
+     * @return string Etag
+     */
+    protected function generateEtag(HttpRequest $request, HttpResponse $response)
+    {
```

Comment:
I propose a slight refactor. Instead of a method called `generateEtag`, how about `getETagGenerator()`? That's the bulk of what this method is actually doing; the actual generation is one line.

`setETag()` would then change the logic to read:

```php
$generator = $this->getETagGenerator();
$etag = new Header\Etag($generator->generate($request, $response));
```

That would also allow reducing the arguments required by this method.